### PR TITLE
Bridge to L2 in key funder

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR allows the key funder to automatically bridge to L2 from L1 when L2 key funder balances are low.

Previously this had to be done manually.

### Drive-by changes

- Upgrade ethers to 5.7 so that we can use the latest optimism SDK

### Backward compatibility

Yes

### Testing
- Ran manually and confirmed that bridging to optimismgoerli and arbitrumgoerli worked

### CODE REVIEW

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 1]
```diff
@@ -268,54 +273,30 @@ class ContextFunder {
...
-    for (const chain of this.chains) {
-      for (const key of keys) {
-        const failure = await this.attemptToFundKey(key, chain);
-        if (failure) {
-          failureOccurred = true;
-        }
-      }
+    for (const [key, chain] of keysAndChains) {
+      const failure = await this.attemptToFundKey(key, chain);
+      failureOccurred = failureOccurred || failure;
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:282]: 🤠

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 2]
```diff
@@ -349,6 +330,51 @@ class ContextFunder {
...
 
+  private async bridgeToL2s(chains: ChainName[]) {
+    const l2s: ChainName[] = [
+      'optimism',
+      'arbitrum',
+      'optimismgoerli',
+      'arbitrumgoerli',
+    ];
+    for (const l2 of l2s) {
+      if (chains.includes(l2)) {
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:341]: this is a bit hard to read
iiuc we want something like chains.filter(isL2).map

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 3]
```diff
@@ -268,54 +273,30 @@ class ContextFunder {
...
-
-    const keys = this.getKeysWithRole(KEY_ROLE_ENUM.Relayer);
-
-    for (const chain of this.chains) {
-      for (const key of keys.filter((k) => k.chainName !== chain)) {
-        const failure = await this.attemptToFundKey(key, chain);
-        if (failure) {
-          failureOccurred = true;
-        }
+  private getKeysAndChains() {
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:287]: dont love this function or the return type but not imp

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 4]
```diff
@@ -357,64 +383,101 @@ class ContextFunder {
...
+      key: keyInfo,
+      receipt,
+      context: this.context,
+      chain,
+    });
+  }
+
+  private async bridgeToL2(l2Chain: ChainName, to: string, amount: BigNumber) {
+    const testnet = l2Chain.includes('goerli');
+    const l1Chain: ChainName = testnet ? 'goerli' : 'ethereum';
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:443]: this is weird control flow
would prefer a static map from L2 chain to L1 chain

Comment (yorhodes) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts:443]: (that throws if L2 has no L1 defined)

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 5]
```diff
@@ -357,64 +383,101 @@ class ContextFunder {
...
+    });
+    let tx;
+    if (l2Chain.includes('optimism')) {
+      const crossChainMessenger = new CrossChainMessenger({
+        l1ChainId: ChainNameToDomainId[l1Chain],
+        l2ChainId: ChainNameToDomainId[l2Chain],
+        l1SignerOrProvider: l1ChainConnection.signer!,
+        l2SignerOrProvider: l2ChainConnection.provider,
+      });
+      tx = crossChainMessenger.depositETH(amount, {
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:465]: not awaited here but is in arbitrum?

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 6]
```diff
@@ -357,64 +383,101 @@ class ContextFunder {
...
+      });
+    } else if (l2Chain.includes('arbitrum')) {
+      const l2Network = await getL2Network(ChainNameToDomainId[l2Chain]);
+      const ethBridger = new EthBridger(l2Network);
+      tx = await ethBridger.deposit({
+        amount,
+        l1Signer: l1ChainConnection.signer!,
+        overrides: l1ChainConnection.overrides,
+      });
+    } else {
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:477]: break these out into their own functions?

File: typescript/infra/src/agents/aws/key.ts [hunk 7]
```diff
@@ -231,6 +231,8 @@ export class AgentAwsKey extends CloudAgentKey {
     if (!keyId) {
       throw Error('Key ID not defined');
     }
+    // @ts-ignore provider.getFeeData type mismatch due to
+    // aws-kms-ethers-signer using ethers 5.6 and this package using ethers 5.7
```
Comment (yorhodes) [inline typescript/infra/src/agents/aws/key.ts:235]: 🤔
should we use npm overrides?
https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts (hunk 5 above)
Comment (asaj) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts]: Good catch

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts (hunk 6 above)
Comment (asaj) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts]: Done

File: typescript/infra/src/agents/aws/key.ts (hunk 7 above)
Comment (asaj) [reply typescript/infra/src/agents/aws/key.ts]: I tried and couldn't get it to work :(

**Review by nambrot — APPROVED** [review]
🙏🙏

**Pull request merged**